### PR TITLE
[WIP]training set generation directory

### PR DESCRIPTION
**Pull request by ohamto**
`run_psi4.py` will loop over a 1B xyz file and calculate reference energies using psi4. It then writes an xyz training set while the outputs are compressed into a zip. Psi4 job control to be added...

### CODE REVIEW

File: psi4_training_set/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#### Training Set Generation
+This directory contains the scripts to read in geometries using pybel and calculates their 1B energies using psi4.
```
Comment (zonca) [inline psi4_training_set/README.md]: add link for `pybel` and also a short description of what it does

File: psi4_training_set/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+`run_psi4.py` uses Open Babel's Pybel wrapper which is a lighter version of the module that allows file input/output etc.
```
Comment (zonca) [inline psi4_training_set/README.md]: > lighter

lighter compared to something else?

File: psi4_training_set/scripts/run_psi4.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
+
+import os, sys
+import psi4
+import pybel as pb
```
Comment (zonca) [inline psi4_training_set/scripts/run_psi4.py]: in `requirements.txt` you ask for `openbabel`, is that the same thing as `pybel`?

File: psi4_training_set/scripts/input [hunk 4]
```diff
@@ -0,0 +1,50 @@
+3
```
Comment (zonca) [inline psi4_training_set/scripts/input]: I recommend to give `input` an extension, I guess there is a standard extension for this kind of files. If not, `.txt` or `.dat` would work.

File: psi4_training_set/scripts/run_psi4.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import shutil
+
+#TODO
+#Things to add later
+#descriptive names for variables etc.
+#reader for qm settings for psi4 like charge and multiplicity
```
Comment (zonca) [inline psi4_training_set/scripts/run_psi4.py]: what is `qm`?

File: psi4_training_set/scripts/run_psi4.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#output storage and compression
+#training set writing during/after calculation
+
+training_set_file = open('./training_set.xyz', 'w')
+
+#read in an xyz file called input and start counting configurations with i_config
+i_config = 0
+for mol in pb.readfile("xyz", "input"):
```
Comment (zonca) [inline psi4_training_set/scripts/run_psi4.py]: is this script for 1 body only? should we call it `run_psi4_1B.py`?

File: psi4_training_set/scripts/run_psi4.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import shutil
+
+#TODO
+#Things to add later
+#descriptive names for variables etc.
+#reader for qm settings for psi4 like charge and multiplicity
+#output storage and compression
+#training set writing during/after calculation
+
+training_set_file = open('./training_set.xyz', 'w')
```
Comment (zonca) [inline psi4_training_set/scripts/run_psi4.py]: it is good practice to use the `with` statement when opening files (see for example http://effbot.org/zone/python-with-statement.htm or google for it)

File: psi4_training_set/README.md (hunk 1 above)
Comment (ohamto) [reply psi4_training_set/README.md]: ok adding a link and description of the two python modules from Open Babel . Thanks

File: psi4_training_set/README.md (hunk 2 above)
Comment (ohamto) [reply psi4_training_set/README.md]: so there are two Python modules available from Open Babel.  There's an `openbabel` module which is, I guess standard, Python binding to the C++ Open Babel library. The `pybel` module has more convenient classes and functions. Both can be used together.

File: psi4_training_set/scripts/run_psi4.py (hunk 3 above)
Comment (ohamto) [reply psi4_training_set/scripts/run_psi4.py]: `pybel` is included in `openbabel` . The Open Babel `pybel` can't be installed by itself. There's a bioconda package called `pybel` which is not what we want to use.

File: psi4_training_set/scripts/run_psi4.py (hunk 5 above)
Comment (ohamto) [reply psi4_training_set/scripts/run_psi4.py]: `quantum mechanics` although that should be changed to something more specific like `job control`...

File: psi4_training_set/scripts/run_psi4.py (hunk 6 above)
Comment (ohamto) [reply psi4_training_set/scripts/run_psi4.py]: right now, yes, it's for 1 body only. I'll rename the file and keep the other n-body scripts separate at the moment.

File: psi4_training_set/scripts/input (hunk 4 above)
Comment (ohamto) [reply psi4_training_set/scripts/input]: ok i'll rename it `input.xyz`

**Comment by zonca** [comment]
@ohamto can you please finalize this so I can merge?

**Comment by ohamto** [comment]
Hi @zonca , the directory and script inside can stand on its own so I think you can merge it. I'll add to the directory and make changes on later PRs.

**Comment by zonca** [comment]
I get this error:

```
(psi4) :~/Paesani/potential_fitting/psi4_training_set/scripts$ python run_psi4_1B.py 
Traceback (most recent call last):
  File "run_psi4_1B.py", line 43, in <module>
    psi4.core.set_output_file('./calculations/'+str(i_config).zfill(4)+'/output.dat')
TypeError: set_output_file(): incompatible function arguments. The following argument types are supported:
    1. (arg0: str, arg1: bool) -> None
```

It looks like `set_output_file` wants a second command line argument, bool.

**Comment by ohamto** [comment]
I added `False` for the second argument. I'm not sure why it was giving an error. The function with and without the bool were working for me.

**Comment by zonca** [comment]
thanks @ohamto now it works, but we need a test.
is `training_set.xyz` already in the format we want to use? can we add a test that compares that to a previous execution?

**Comment by ohamto** [comment]
`training_set.xyz` is in the format of our previous 1B training sets. I'll have to check for the formats of the higher-body training sets but right now I was thinking of using `input.xyz` and `training_set.xyz` for testing.

**Comment by zonca** [comment]
yes @ohamto what would be perfect

**Comment by ohamto** [comment]
ok I added a tests directory with a `test_run_psi4_1B.sh` that compares the generated training sets

**Comment by zonca** [comment]
thanks @ohamto , I think this is good to merge, @agoetz  do you have any other feedback?

**Comment by zonca** [comment]
@ohamto your latest test fails, there is too much precision in those comparisons, it fails with:

```
47c47
< -75.58547595579591
---
> -75.58547595571959
run_psi4_1b.py error!
```

is there a way to limit the number of digits? or to load the file into python and compare numbers instead of strings?

**Comment by ohamto** [comment]
@zonca I can limit the number of digits for now through PSI4

**Comment by ohamto** [comment]
@zonca Ok I limited the reference energies to 8 decimal places and replaced the expected training set in testing

**Comment by agoetz** [comment]
Great! I think this is a good start. We will have to extend and refactor the code but I think for now we can use it as a basis.

**Comment by zonca** [comment]
ok, I'll test again on my machine and then merge

**Comment by zonca** [comment]
it passes on my machine, good!

**Comment by agoetz** [comment]
Test passes on my macbook. Next steps:

- Settings for psi4 calculations through input file
- Calculator object that stores psi4 settings and has method that executes psi4

@zonca @ohamto The current setup goes sequentially through the list of structures, executes psi4 for each, and adds to the training set file. Do you have a suggestion how we can modify this for parallel execution / distributed computing?

**Pull request merged**